### PR TITLE
SNOW-783542 Support dataframe.join for local testing

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-783542

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

WIP: Enabled 12/69 integ tests that hit dataframe.join

- [x] support inner join
- [x] support outer join
- [x] support session.range (for porting tests)
- [x] support natural join
- [x] support cross join
- [x] support left outer join
- [x] support right outer join
- [x] support left semi join
- [x] support left anti join
- [ ] support expr id to col name map

### CODE REVIEW

File: src/snowflake/snowpark/dataframe.py [hunk 1]
```diff
@@ -504,7 +508,7 @@ def __init__(
         self._plan = self._session._analyzer.resolve(plan)
         if isinstance(plan, (SelectStatement, MockSelectStatement)):
             self._select_statement = plan
-            plan.expr_to_alias.update(self._plan.expr_to_alias)
+            # plan.expr_to_alias.update(self._plan.expr_to_alias)
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/dataframe.py:511]: This needs to be uncommented.

File: src/snowflake/snowpark/dataframe.py [hunk 2]
```diff
@@ -1897,10 +1904,22 @@ def natural_join(
             None,
         )
         if self._select_statement:
-            select_plan = SelectStatement(
-                from_=SelectSnowflakePlan(join_plan, analyzer=self._session._analyzer),
-                analyzer=self._session._analyzer,
-            )
+            if isinstance(self._select_statement, MockSelectable):
+                select_plan = MockSelectStatement(
+                    projection=[Star([])],
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/dataframe.py:1910]: Maybe projection here could be default to `Star([])`

File: src/snowflake/snowpark/dataframe.py [hunk 3]
```diff
@@ -3189,28 +3232,28 @@ def na(self) -> DataFrameNaFunctions:
...
-        Args:
-            cols: The names of columns whose basic statistics are computed.
+                Computes basic statistics for numeric columns, which includes
+                ``count``, ``mean``, ``stddev``, ``min``, and ``max``. If no columns
+                are provided, this function computes statistics for all numerical or
+                string columns. Non-numeric and non-string columns will be ignored
+                when calling this method.
+
+                Example::
+        import snowflake.snowpark.mock.mock_select_statement            >>> df = session.create_dataframe([[1, 2], [3, 4]], schema=["a", "b"])
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/dataframe.py:3242]: This is probably a Pycharm refactor mistake. Reverting...

File: src/snowflake/snowpark/mock/mock_datastore.py [hunk 4]
```diff
@@ -1,43 +0,0 @@
-#
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/mock_datastore.py:1]: Deleted this file because it's not being referenced.

File: src/snowflake/snowpark/mock/mock_plan.py [hunk 5]
```diff
@@ -254,6 +279,149 @@ def execute_mock_plan(plan: MockExecutionPlan) -> TableEmulator:
...
+                ].dropna()
+            elif "ANTI" in source_plan.join_type.sql:  # left anti
+                result_df = left[
+                    ~(
+                        left.apply(tuple, 1).isin(
+                            result_df[condition][left.columns].apply(tuple, 1)
+                        )
+                    )
+                ].dropna()
+            elif "LEFT" in source_plan.join_type.sql:  # left outer join
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/mock_plan.py:384]: The left outer, righter outer and full outer have some duplicate code. Refactoring...

File: src/snowflake/snowpark/dataframe.py [hunk 6]
```diff
@@ -1897,10 +1905,22 @@ def natural_join(
...
+                    ),
+                    analyzer=self._session._analyzer,
+                )
+            else:
+                select_plan = SelectStatement(
+                    from_=SelectSnowflakePlan(
+                        join_plan, analyzer=self._session._analyzer
+                    ),
+                    analyzer=self._session._analyzer,
+                )
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/dataframe.py:1923]: do you think we can move the logic into mock analyzer, into those `create_*` methods?

I was previously discussing with Yijun that we want to introduce as few changes in `dataframe` as possible to keep the mocking related feature modularized/isolated.

File: src/snowflake/snowpark/dataframe.py [hunk 7]
```diff
@@ -701,7 +706,10 @@ def to_local_iterator(
         )
 
     def __copy__(self) -> "DataFrame":
-        return DataFrame(self._session, copy.copy(self._plan))
+        if isinstance(self._select_statement, MockSelectStatement):
+            return DataFrame(self._session, copy.copy(self._select_statement))
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/dataframe.py:710]: why copying `self._plan` does not work for mock select statement?

File: src/snowflake/snowpark/mock/mock_plan.py [hunk 8]
```diff
@@ -1,388 +0,0 @@
-#
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/mock_plan.py:1]: Moved all code to `mock_select_statement` to resolve the circular dependency that `MockExecutionPlan` references classes defined in `mock_select_statement` and `mock_select_statement` relies on executing `MockExecutionPlan` to retrieve attributes.

File: src/snowflake/snowpark/dataframe.py [hunk 9]
```diff
@@ -2315,6 +2335,17 @@ def _join_dataframes(
                 None,
             )
             if self._select_statement:
+                if isinstance(self._select_statement, MockSelectStatement):
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/dataframe.py:2338]: same as above, do you think we can move the logic to mock analyzer?

File: src/snowflake/snowpark/dataframe.py (hunk 7 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/dataframe.py:710]: Because in `dataframe.py`, the code checks for `self._select_statement` to determine if the SQL simplifier is enabled when this dataframe is constructed. In the current Snowpark Python, copying a dataframe would lose this information and render all following operations non-SQL-simplified. This is a known issue that Yijun and I had tried to tackle, but making `copy` works for SQL simplifier codepath involves changing `expr_to_alias`, which I spent days and gave up on 😢  
In short: `copy` will make a dataframe no longer SQL-simplifiable. This is an existing issue.

File: tests/integ/scala/test_dataframe_suite.py [hunk 10]
```diff
@@ -913,7 +913,8 @@ def test_sample_on_union(session):
     )
 
 
-@pytest.mark.localtest
+# TODO: Re-enable local testing, currently this fails at agg
+# @pytest.mark.localtest
```
Comment (sfc-gh-stan) [inline tests/integ/scala/test_dataframe_suite.py:917]: I'm hoping rebasing on Adam's aggregation PR would resolve this. If not, I will make sure this test gets re-enabled before merging this PR.

File: tests/integ/scala/test_dataframe_suite.py [hunk 11]
```diff
@@ -179,7 +179,7 @@ def test_adjust_column_width_of_show(session):
 
 
 @pytest.mark.localtest
-def test_show_with_null_data(session):
+def oh_null_data(session):
```
Comment (sfc-gh-stan) [inline tests/integ/scala/test_dataframe_suite.py:182]: Will revert.

File: tests/integ/scala/test_dataframe_suite.py [hunk 12]
```diff
@@ -240,7 +240,7 @@ def test_show(session):
     # make sure show runs with sql
     session.sql("show tables").show()
 
-    session.sql("drop table if exists test_table_123").show()
```
Comment (sfc-gh-stan) [inline tests/integ/scala/test_dataframe_suite.py:243]: Will revert.

File: src/snowflake/snowpark/mock/mock_select_statement.py [hunk 13]
```diff
@@ -62,17 +108,19 @@ def __init__(
...
         self._column_states: Optional[ColumnStateDict] = None
-        self.expr_to_alias = {}
         self._execution_plan: Optional[SnowflakePlan] = None
         self._attributes = None
 
     @property
     def execution_plan(self):
+        """Convert to a SnowflakePlan"""
+        if self._execution_plan is None:
+            self._execution_plan = MockExecutionPlan(self, self.analyzer.session)
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/mock_select_statement.py:118]: is this the reason for the circular import?
I'm wondering what's the case that `self._execution_plan` could be passed as None and we need to wrap it here.

**Review by sfc-gh-aling — APPROVED** [review]

**Review by sfc-gh-aling — APPROVED** [review]
